### PR TITLE
BAEL-3852 - A Guide to Foreign Memory Access API in Java 14

### PR DESCRIPTION
**Pull request by musibs**

### CODE REVIEW

File: core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+import static org.hamcrest.core.Is.*;
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.*;
+import static org.junit.Assert.assertThat;
+
+public class ForeignMemoryUnitTest {
+
+    @Test
+    public void whenValueIsSetThenAccessTheValue(){
```
Comment (ashleyfrieze) [inline core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java:19]: Missing ` ` between `{}` and `{`

File: core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            }
+
+            for(int i=0; i<25; i++){
+                assertThat(varHandle.get(base.addOffset((i*4))), is(i));
+            }
+        }
+    }
+
+    @Test
+    public void testMemoryLayout(){
```
Comment (ashleyfrieze) [inline core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java:43]: Unit tests with names that start `test` often don't explain themselves well. Let's stick with the `given_when_then` style. Maybe `whenSetValuesWithMemoryLayout_thenTheyCanBeRetrieved`

File: core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            }
+
+            for(long i=0; i<sequenceLayout.elementCount().getAsLong(); i++){
+                assertThat(varHandle.get(base, i), is(i));
+            }
+        }
+    }
+
+    @Test
+    public void testSlicingOfMemorySegment(){
```
Comment (ashleyfrieze) [inline core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java:59]: Similar test naming issue here.

File: core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java (hunk 1 above)
Comment (musibs) [reply core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java]: Added space between `()` and `{` in all unit test methods.

File: core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java (hunk 2 above)
Comment (musibs) [reply core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java]: Updated all the unit test method names with the `given_when_then`  style.

File: core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java (hunk 3 above)
Comment (musibs) [reply core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java]: Updated all the unit test method names with the `given_when_then` style.

File: core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import static org.junit.Assert.*;
+import static org.junit.Assert.assertThat;
+
+public class ForeignMemoryUnitTest {
+
+    @Test
+    public void whenAValueIsSet_thenAccessTheValue() {
+        long value = 10;
+        MemoryAddress memoryAddress = MemorySegment
+                .allocateNative(8).baseAddress();
```
Comment (ashleyfrieze) [inline core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java:22]: FWIW, I find this harder to read than if it were:

```java
        MemoryAddress memoryAddress = 
            MemorySegment.allocateNative(8)
                .baseAddress();
```

The reason is that the hanging `MemorySegment` on the original, is lost from my memory when I read the following line

File: core-java-modules/core-java-perf/src/main/java/com/baeldung/threaddump/ThreadDump.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    private static void threadDump(boolean lockedMonitors, boolean lockedSynchronizers) throws IOException {
+        Path threadDumpFile = Paths.get("ThreadDump.txt");
+        
+        StringBuffer threadDump = new StringBuffer(System.lineSeparator());
+        ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
+        for(ThreadInfo threadInfo : threadMXBean.dumpAllThreads(lockedMonitors, lockedSynchronizers)) {
+            threadDump.append(threadInfo.toString());
+        }
+        Files.write(threadDumpFile, threadDump.toString().getBytes());
+    }
```
Comment (ashleyfrieze) [inline core-java-modules/core-java-perf/src/main/java/com/baeldung/threaddump/ThreadDump.java:26]: Why is this here?

Comment (musibs) [reply core-java-modules/core-java-perf/src/main/java/com/baeldung/threaddump/ThreadDump.java]: This commit has accidentally become part of this pull request. It was for my previous article. I have deleted the file and committed the changes.

File: core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java (hunk 4 above)
Comment (musibs) [reply core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java]: I have aligned the code snippets as per your suggestion.

File: core-java-modules/core-java/src/main/java/com/baeldung/java/hexagonal/DefaultUserService.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package com.baeldung.java.hexagonal;
+
+import java.util.Collection;
+import java.util.Optional;
+
+/**
+ *  Default implementation of user service 
+ */
+public class DefaultUserService implements UserService{
```
Comment (sinsuren) [inline core-java-modules/core-java/src/main/java/com/baeldung/java/hexagonal/DefaultUserService.java:9]: indentation.

File: core-java-modules/core-java/src/main/java/com/baeldung/java/hexagonal/InMemoryUserRepositoryImpl.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
+package com.baeldung.java.hexagonal;
+
+/**
+ * An in-memory implementation of the UserRepository
+ */
+import java.util.Collection;
+import java.util.Optional;
+
+public class InMemoryUserRepositoryImpl implements UserRepository{
```
Comment (sinsuren) [inline core-java-modules/core-java/src/main/java/com/baeldung/java/hexagonal/InMemoryUserRepositoryImpl.java:9]: UserRepository {

File: core-java-modules/core-java/src/main/java/com/baeldung/java/hexagonal/InMemoryUserRepositoryImpl.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+	@Override
+	public User save(User user) {
+		return null;
```
Comment (sinsuren) [inline core-java-modules/core-java/src/main/java/com/baeldung/java/hexagonal/InMemoryUserRepositoryImpl.java:13]: space indentation is also not uniform across files.

File: core-java-modules/core-java/src/main/java/com/baeldung/java/hexagonal/DefaultUserService.java (hunk 6 above)
Comment (ashleyfrieze) [reply core-java-modules/core-java/src/main/java/com/baeldung/java/hexagonal/DefaultUserService.java:9]: These changes are tab indented and should not be in this PR. Please revert them.

Comment (musibs) [reply core-java-modules/core-java/src/main/java/com/baeldung/java/hexagonal/DefaultUserService.java]: @sinsuren and @ashleyfrieze 

Thanks for the review. These files are part of my first demo article and accidentally part of this PR. I have now removed the complete package. I'll be careful and verify the commits part of the PR.

File: core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+import static org.junit.Assert.*;
+import static org.junit.Assert.assertThat;
+
+public class ForeignMemoryUnitTest {
+
+    @Test
+    public void whenAValueIsSet_thenAccessTheValue() {
+        long value = 10;
+        MemoryAddress memoryAddress =
+                MemorySegment.allocateNative(8)
```
Comment (KevinGilmore) [inline core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java:22]: Indentation guidelines prescribe a 4-space indent for nested blocks and 2-space indent for continuation lines. Please adjust the indentation in this file accordingly.

Comment (musibs) [reply core-java-modules/core-java-14/src/test/java/com/baeldung/java14/foreign/api/ForeignMemoryUnitTest.java]: Thanks for the review and the feedback. I've incorporated the changes as asked. Kindly review,
